lib/io/hardware_can.cpp

```diff
@@ -0,0 +1,127 @@
+#include "hardware_can.hpp"
+
+#include <stdio.h>
+#include <unistd.h>
+
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+
+namespace hyped::io {
+
+std::optional<HardwareCan> HardwareCan::create(core::ILogger &logger,
+                                               const std::string &can_network_interface)
+{
+  std::int16_t socket_id = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+  if (socket_id < 0) {
+    logger.log(core::LogLevel::kFatal, "Unable to open CAN socket");
+    return std::nullopt;
+  }
+  const std::uint8_t interface_index = if_nametoindex(can_network_interface.c_str());
+  if (!interface_index) {
+    logger.log(core::LogLevel::kFatal, "Unable to find CAN1 network interface");
+    close(socket_id);
+    return std::nullopt;
+  }
+  const sockaddr_can socket_address = {AF_CAN, interface_index};
+  const int bind_status
+    = bind(socket_id, reinterpret_cast<const sockaddr *>(&socket_address), sizeof(socket_address));
+  if (bind_status < 0) {
+    logger.log(core::LogLevel::kFatal, "Unable to bind CAN socket");
+    close(socket_id);
```

Comment:
you probably want to do this in a deconstructor, too.

but beware, there's some subtlety here: you return a `HardwareCan`; this means that deallocation happens all the time, whenever the object is copied somewhere; things might also get out of sync between the copies in other ways

solution: return a shared pointer; of course this doesn't quite prevent anybody from copying the object around but it conveys the intention. we could make this impossible on a type level but i don't believe that that's worth the hassle